src/KeyManager.sol

```diff
@@ -33,62 +38,103 @@ abstract contract KeyManagerBase {
     // master private key
     function xPriv() internal virtual returns (bytes32);
 
-    // Key derivation for encryption
+    function _derivedPriv(address a, uint32 index) internal virtual returns (bytes32);
 
-    // Any contract in confidential mode can request a
-    // hardened derived key
-    function _derivedPriv(address a) private returns (bytes32) {
-        return keccak256(abi.encodePacked(a, xPriv()));
-    }
-
-    function derivedPriv() public returns (bytes32) {
-        return _derivedPriv(msg.sender);
-    }
+    // Key derivation for encryption
 
     // Because we are using hardened derivation, for each
     // contract we will need someone to sign it off chain
-    mapping(address => bytes) public derivedPub;
+    mapping(bytes32 => bytes) internal _derivedPub;
 
     function onchain_DeriveKey(
         address a,
         bytes memory dPub,
         // Signature from a valid kettle
-        bytes memory sig
+        bytes memory sig,
+        uint32 index
     ) public {
-        bytes32 digest = keccak256(abi.encodePacked(a, dPub));
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
         require(Secp256k1.verify(xPub, digest, sig));
-        derivedPub[a] = dPub;
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        _derivedPub[key] = dPub;
     }
 
-    function offchain_DeriveKey(address a) public returns (bytes memory dPub, bytes memory sig) {
-        bytes32 dPriv = _derivedPriv(a);
+    function offchain_DeriveKey(address a, uint32 index) public returns (bytes memory dPub, bytes memory sig) {
+        bytes32 dPriv = _derivedPriv(a, index);
         dPub = PKE.derivePubKey(dPriv);
-        bytes32 digest = keccak256(abi.encodePacked(a, dPub));
+        bytes32 digest = keccak256(abi.encodePacked(a, index, dPub));
         sig = Secp256k1.sign(uint256(xPriv()), digest);
         require(Secp256k1.verify(xPub, digest, sig));
     }
 }
 
 contract KeyManager_v0 is KeyManagerBase {
     IAndromeda public Suave;
+    BIP32 public bip32;
+
 
     constructor(address _Suave) {
         Suave = IAndromeda(_Suave);
+        bip32 = new BIP32(Suave);
+    }
+
+    function derivedPub(address a, uint32 index) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, index));
+        return _derivedPub[key];
+    }
+
+    function derivedPub(address a) public view returns (bytes memory) {
+        bytes32 key = keccak256(abi.encodePacked(a, uint32(0)));
```

Comment:
could be more clear if you did `return derivedPub(a, 0)`

Reply:
This leads to compiler error

> Cannot perform packed encoding for a literal. Please convert it to an explicit type first.